features/steps/thens/relations.py

```diff
@@ -131,6 +126,157 @@ def step_impl(context, related, relating, other_entity, condition):
                         yield(err.RuleSuccessInst(True, inst))
 
 
+@then('The IfcPropertySet Name attribute value must use predefined values according to the {table} table')
+@then('The IfcPropertySet must be assigned according to the property set definitions table {table}')
+@then('Each associated IfcProperty must be named according to the property set definitions table {table}')
+@then('Each associated IfcProperty must be of type according to the property set definitions table {table}')
+@then('Each associated IfcProperty value must be of data type according to the property set definitions table {table}')
+@err.handle_errors
+def step_impl(context, table):
+    if getattr(context, 'applicable', True):
+        tbl_path = system.get_abs_path(f"resources/property_set_definitions/{table}")
+        tbl = system.get_csv(tbl_path, return_type='dict')
+        property_set_definitons = {}
+        for d in tbl:
+            property_set_definitons[d['property_set_name']] = d
+
+        def establish_accepted_pset_values(name, property_set_definitons):
+            def make_obj(s):
+                if s:
+                    return json.loads(s.replace("'", '"'))
+                else:
+                    return ''
+
+            try:
+                property_set_attr = property_set_definitons[name]
+            except KeyError:  # Pset_ not found in template
+                property_set_attr = ''
+                return property_set_attr
+
+            accepted_values = {}
+            accepted_values['template_type'] = property_set_attr.get('template_type', '')
+
+            accepted_values['property_names'] = []
+            accepted_values['property_types'] = []
+            accepted_values['data_types'] = []
+
+            for property_def in make_obj(property_set_attr['property_definitions']):
+                accepted_values['property_names'].append(property_def['property_name'])
+                accepted_values['property_types'].append(property_def['property_type'])
+                accepted_values['data_types'].append(property_def['data_type'])
+
+            accepted_values['applicable_entities'] = make_obj(property_set_attr['applicable_entities'])
+
+            try:
+                accepted_values['applicable_type_values'] = make_obj(property_set_attr.get('applicable_type_value', '')).split(',')
+            except json.decoder.JSONDecodeError:
```

Comment:
Can you maybe explain what happens here? With `json.decoder.JSONDecodeError:`

Reply:
Good question as I had to remind myself why it's there :) 

The reason was that I was trying to make a python object from string like: `"['IfcActor', 'IfcBuilding', 'IfcSite']"`. If a string was '', json.loads() produced an error. Perhaps it was cleaner to check if a string was empty instead of the try/except. 

I've checked though, and for `applicable_type_values `there's no need to use json at all, as it's always in a nice string format, like: `"IfcActor,IfcBuilding,IfcSite"`. 

I've changed those lines removing this unnecessary redundancy. Thank you for a very valid point.